File: js/Samples/PathFinder_Refactored.js
class PathFinder_Refactored {
    *getPathToTarget(maze, start, target) {
        let source = start;
        const prev = {};

        while (source) {
            if (prev[source.toString()] === undefined) {
                prev[source.toString()] = this.getNextStepToTarget(source, target);
            }

            yield prev[source.toString()];
            source = prev[source.toString()];
        }
    }

    getNextStepToTarget(maze, source, target) {
        const queue = [];
        const used = [];
        queue.push(target);
        used.push(target);

        while (queue.length) {
            const p = queue.shift();
            for (const neighbour of this.getNeighbours(p, maze)) {
                const isPointVisited = contains(used, neighbour);
                if (isPointVisited) continue;
                if (comparePoints(neighbour, source)) {
                    return p;
                }
                queue.push(neighbour);
                if (!isPointVisited) {
                    used.push(neighbour);
                }
            }
        }

        return source;
    }

    generateRandomWorld() {
        return {
            insideMaze: (location) => {},
            isFree: (location) => {}
        };
    }

    getNeighbours(from, maze) {
        return [{x: 1, y: 0}, {x: -1, y: 0}, {x: 0, y: 1}, {x: 0, y: -1}]
            .map(shift => ({ x: shift.x + from.x, y: shift.y + from.y }))
            .filter(maze.insideMaze)
            .filter(maze.isFree);
    }
}

function contains(arr, source) {
    return arr.some(compareTo(source));
}

function comparePoints (point1, point2) {
    return point1.x === point2.x && point1.y === point2.y;
}

function compareTo(point1) {
    return (point2) => comparePoints(point1, point2);
}

File: python/Samples/path_finder_refactored.py
from dataclasses import dataclass
from typing import List


@dataclass
class Point:
    x: int
    y: int

    def __add__(self, other):
        return Point(self.x + other.x, self.y + other.y)

    def __str__(self):
        return f"({self.x}, {self.y})"


@dataclass
class Maze:
    inside_maze: list
    is_free: list


class PathFinder:
    @staticmethod
    def generate_random_world():
        max_x, max_y = 4, 4
        maze_points = [Point(x, y) for x in range(max_x) for y in range(max_y)]
        return maze_points, maze_points

    @staticmethod
    def get_path_to_target(maze: Maze, start: Point, target: Point):
        prev = {}
        source = start
        while source:
            if not prev.get(str(source)):
                prev[str(source)] = PathFinder.get_next_step_to_target(maze, source, target)
            if prev[str(source)]:
                yield prev[str(source)]
            source = prev[str(source)]

    @staticmethod
    def get_next_step_to_target(maze: Maze, source: Point, target: Point):
        queue = [target, ]
        used = [target, ]
        while len(queue) > 0:
            point = queue.pop(0)
            for neighbour in PathFinder.get_neighbours(point, maze):
                if neighbour in used:
                    continue
                if neighbour == source:
                    return point
                queue.append(neighbour)
                used.append(neighbour)

        return None

    @staticmethod
    def get_neighbours(point: Point, maze: Maze) -> List[Point]:
        offsets = [Point(0, 1), Point(1, 0), Point(-1, 0), Point(0, -1)]
        points = [point + offset for offset in offsets]
        return [
            point
            for point in points
            if point in maze.inside_maze and point in maze.is_free
        ]
